### ISSUE TITLE
 upgrade trezor-connect dependency to latest

### ISSUE THREAD

**Issue by prusnak** [MetaMask/metamask-extension#8349]
Please upgrade the trezor-connect[1] dependency of your application to
latest (8.1.5 atm), otherwise your users won't be able to enter the
passphrase on the device.

Thanks!

[1] https://www.npmjs.com/package/trezor-connect

**Comment by whymarrh** [comment]
We are currently using 7.0.3. The set of breaking changes in v8 is here: [`CHANGELOG.md`](https://github.com/trezor/connect/blob/develop/CHANGELOG.md#800)

**Comment by jcbit** [comment]
bump

**Comment by netoracle** [comment]
When will the new version be released which addresses the trezor connect issue prusnak mentioned above?

Thanks

**Comment by blockomoco** [comment]
any update to this ?

**Comment by whymarrh** [comment]
> otherwise your users won't be able to enter the passphrase on the device.

What specifically doesn't work? And under what circumstances doesn't it work? I don't understand and this issue doesn't have specific details.

**Comment by dpazdan** [comment]
further details mentioned here: https://www.reddit.com/r/TREZOR/comments/hhs9k1/what_is_this_error_chrome_8304103116_official/

**Comment by whymarrh** [comment]
We'll hopefully update our Trezor logic in 8.1 but in the meantime I do believe it still _works_.

**Comment by epop-cs** [comment]
Hello @whymarrh,

https://github.com/MetaMask/metamask-extension/issues/8866 was closed and I was asked to comment here. I do not believe these two tickets represent the same issue. This one is referring to the Trezor Connect API version 8.1.7, while the issue the user reported to support refers to the connection he/she is attempting to make to their physical Trezor wallet through the Metamask UI. It clearly mentions the 2.3.1 version in the screenshots posted inside the initial issue.
![switccheo](https://user-images.githubusercontent.com/58727092/86017184-6a1f8780-ba2c-11ea-869d-022781845a7b.png)


Can you please advise?

Regards,

Emil

**Comment by blockomoco** [comment]
Currently (with the outdated Trezor integration) once needs to type the Trezor typephrase in the browser which is **not secure**. A major security aspect of hardware wallets is to avoid typing secrets into potentially compromised devices (like a computer with internet access). Also, the passphrase needs to be typed in to sign _every_ single transaction, which is bad UX and fairly annoying to be frank.

According to the Trezor team (I sanity checked with them), if you upgrade the Trezor integration these 2 issues should be resolved so that:
a) the passphrase can be typed in on the Trezor device instead of typing it in the browser
b) one would not need to type the passphrase every time in order to sign transactions

**Comment by ncsolar** [comment]
![Trezor-error](https://user-images.githubusercontent.com/8175508/86021101-57865d80-b9ee-11ea-871e-59b7f5c92269.png)

Same warning here, Chrome Version 83.0.4103.116 (Official Build) (64-bit)

**Comment by netoracle** [comment]
Have a look at how Myetherwallet does it. They have it setup correctly, so
that you can type your passphrase on your device, rather than a keyboard.

**Comment by whymarrh** [comment]
@epop-cs as @ncsolar and @dpazdan has shared, these do all seem to be related.

@blockomoco thanks for the details, that is clearer.

We'll be working to get this in 8.1, the version after next, and will update here with progress. We understand that this isn't a smooth process right now, we're working to fix it.

**Comment by blockomoco** [comment]
Any update when this will be rolled out ?

**Comment by whymarrh** [comment]
No updates right now

**Comment by HeroHann** [comment]
As it seems the problem still exists. (I guess Brave Wallet is using MetaMask?!. -> So they are having the same problem.)

It is NOT: "not going so smooth"
It IS: "not usable for all Trezor users as this is a security risk"

Please fix this asap.

**Comment by blockomoco** [comment]
+1 this is a serious security risk since the password should only be typed on the Trezor device directly and never on a computer with internet access.

Please prioritize this fix. The right thing to do here is fixing and releasing the fix ASAP.

@whymarrh

**Comment by steelbom** [comment]
I also have to second this - an urgent fix would be great.

**The issues I'm having:**
1. I am unable to move funds from my Trezor (Model T) wallet through metamask.
2. I am unable to transact with that wallet in any platform that utilises metamask (uniswap, etc.)
3. I am forced to enter my passphrase in plain text, rather than having the option to enter it on the device.

About two weeks ago it was working fine on the older version of the Trezor firmware that I was using. I updated to 2.3.2 and I have also updated Trezor Connect as well, but now this issue occurs. This is in Vivaldi, on Windows 10 Pro 2004, with adblock disabled.

Error: [link](https://gyazo.com/2cb576b512ec43ec18132ef4de4ae6f3)

**Comment by danfinlay** [comment]
So it sounds like one of the breaking changes Trezor introduced was related to how passwords are entered when connecting with MetaMask?

Does that explain [this user](https://www.reddit.com/r/Metamask/comments/iamyon/is_there_a_metamask_discord_channel/g222x1j/?utm_source=share&utm_medium=ios_app&utm_name=iossmf)'s experience? They claim they are unable to use BIP39 password derived accounts at all anymore.

**Comment by Prince-Fish** [comment]
Bump - this issue effectively breaks secure use of Trezor with Metamask.  Due to its ubiquity, Trezor are forced to rely on Metamask to interact with web3 applications.  Current version of Metamask effectively breaks possibility of secure Trezor use.

**Comment by blockomoco** [comment]
What is the status of this ? It is shocking that this is still unresolved.

Security issues must have the highest priority and urgency. Especially for hardware wallets where people may store significant amount of funds.

Please take action.

**Comment by HeroHann** [comment]
@PatrykLucka @jennypollack Please update us!

I am no dev, but is it that hard?

**Comment by danfinlay** [comment]
We've begun work on this issue, but have encountered some problems, documenting here to share with Trezor:

 The problem is that when trying to connect Trezor, we are redirected to https://connect.trezor.io/8/popup.html#loading and it’s getting stuck on loading screen (exactly the same way when we manually open this site), there’s nothing more we can do. I have latest firmware on my model T and latest trezor-bridge versions.

1. I tried updating trezor-connect to many different 8.x.x versions (including extended one), but none of them works - extended versions runs some unit tests, but should resolve if we have a public key test is getting timeout from TrezorConnecty.getPublicKey() function. Also tried using .ethereumGetPublicKey() , but it behaves in the same way
2. Same function works fine using https://trezor.github.io/connect-explorer/#/
3. I tested same function using their example from https://github.com/trezor/connect/tree/develop/examples/webextension and it also works so it looks like the problem of implementation, not firmware/bridge versions
4. I also checked how MEW has this implemented and it also looks almost the same :man-shrugging:

Any tips or suggestions will be appreciated.

**Comment by blockomoco** [comment]
Thx for working on this.

@prusnak can you folks please help. Thx

**Comment by prusnak** [comment]
@danfinlay Show us the code - maybe push this into a branch somewhere?

**Comment by Prince-Fish** [comment]
So, I went to make some trades today, and *surprise* this isn't fixed.

What's the update on timing for this critical security vulnerability / functionality break that was disclosed half a year ago?

**Comment by gruvin** [comment]
LOL sigh Crazy! :'( WTF?

**Comment by steelbom** [comment]
Any update on this?

**Comment by HD125823** [comment]
@danfinlay a funny suggestion but have you tried testing trezor.io/8 after clearing the cache and/or switching of blockers? I sometimes get stuck on the loading page as well when trying to connect the trezor with metamask. Usually clearing the cache does the trick for me...
https://trezor.io/support/technical/stuck-at-loading/

Hope you guys manage to fix this soon :/

**Comment by chonghe1** [comment]
Still waiting the issue to be fixed

**Comment by mkalen** [comment]
MetaMask does not work anymore with Trezor Model T and the latest firmware (2.3.3), p.find is undefined - see screenshot 
![bild](https://user-images.githubusercontent.com/2629399/95022058-0f25ae00-0675-11eb-9f9c-0eb2f2c7444e.png)
@danfinlay Did you see the offering from prusnak to review code e.g. on a branch? Thanks.

**Comment by blockomoco** [comment]
@prusnak @danfinlay status ?

**Comment by Prince-Fish** [comment]
I understand that this is one issue of many.  To support fixing this issue, I will donate 0.1 Eth to whoever gets this done by Oct 12 '20

I look forward to sending that amount from my Trezor.

**Comment by HD125823** [comment]
I am not familiar with web dev, but it seems to me that we only need to update dependencies in `yarn.lock`.

Specifically these two blocks:

(here we need to change to 8.1.15)
```
trezor-connect@^7.0.1:
  version "7.0.3"
  resolved "https://registry.yarnpkg.com/trezor-connect/-/trezor-connect-7.0.3.tgz#70c4bc26c0966e794fc280a12c1acc9fef88864f"
  integrity sha512-1Y1ajCDF8dC5d2yrCUmVkNqXeOlucamQ6j6Ko7kaqNdge3g9KZ+O48jUwP/eGzei8oUvPZUHd7o4OhDHTlpLCw==
  dependencies:
    "@babel/runtime" "^7.3.1"
    events "^3.0.0"
    whatwg-fetch "^3.0.0"
````
and 

(in here we need to change the last line to `trezor-connect "^8.1.15"`)
```
eth-trezor-keyring@^0.4.0:
  version "0.4.0"
  resolved "https://registry.yarnpkg.com/eth-trezor-keyring/-/eth-trezor-keyring-0.4.0.tgz#f59c210f95aaf3d7321ae69d2b87a3b8db96a828"
  integrity sha512-7F+C1ztxZStLzmG6r/2/MxjSuxw0aU9T26unJ03fQslktKG9izP+dU2IAJUnWxnyej2ZkfcgcH9M1t32LFbK2A==
  dependencies:
    eth-sig-util "^1.4.2"
    ethereumjs-tx "^1.3.4"
    ethereumjs-util "^5.1.5"
    events "^2.0.0"
    hdkey "0.8.0"
    trezor-connect "^7.0.1"
```

I will test this later myself and see if it works.

**Comment by mkalen** [comment]
> I am not familiar with web dev, but it seems to me that we only need to update dependencies in `yarn.lock`.

Unfortunately that's not it. If you read up on the history of this issue you see that there are well documented API breaking changes on the Tezor side and this means more work for integrators (semantic versioning will signal this is the increased major version number, here from 7 to 8, that's also natural as API:s progress).

But you can also see that MetaMask side has aldready upgraded calling code and have it ready for testing. The blocker is that it didn't work and Trezor then offered to help with code review. Solution is within reach.

(I'm not in either team BTW, a MetaMask-user with a Trezor - like you.)

**Comment by HD125823** [comment]
@mkalen I see, too bad then :/ Hopefully, it won't take much longer until this is fixed. 
I read your previous message regarding a failed transaction with model T (2.3.3) via metamask, its strange cause I still can transact without failures, though it is displaying the same warning posted by ncsolar  https://github.com/MetaMask/metamask-extension/issues/8349#issuecomment-651173909

**Comment by sdtsui** [comment]
Hi - previous contributor to MetaMask here. 

I've been getting a lot of questions about Trezor connectivity issues from friends and former clients.  While this note may not apply to you -- it's a simple fix if it actually does.  I debated opening a new issue or contacting a team member directly, but decided on commenting here since given how long this issue has been open, doing it this way is most likely to reach/help the correct people while being unlikely to confuse others.

I don't have time enough to go through the full thread, but what I've noticed is that:
(i) some users are getting errors like @mkalen 's - `p.find(...)` undefined with the latest firmware.
(ii) some users are getting what amounts to a timeout - in a web wallet, Trezor does not get recognized and is the user is never prompted to export their public key, or enter their PIN.  This is closer to what @danfinlay is describing on 09/03.

For users who are getting #1 - this message **does not** apply. I don't know how to help you yet.
For users who are getting #2, especially if you can get BTC to display, or get it working on other interfaces (like MEW). it might be a quick fix for you. It is some sort of browserside cache issue. See below:

https://www.reddit.com/r/TREZOR/comments/j5awtq/my_trezor_t_is_synchronized_with_exodus_i_have/
https://np.reddit.com/r/ExodusWallet/comments/j5axtu/my_trezor_t_is_synchronized_with_exodus_i_have/g7qwm7g/

For some reason, simply clearing the history in chrome did not work.    All I had to do for these users was **install brave and set up wallet with Brave's internal fork of MetaMask** - the Trezor was recognized right away.
This applies to both Trezor Model T and One.  This worked despite firmware being out of date.  I think this is because it's a clean install, and the Trezor knows to start a new session whereas with the old browser it's expecting an old session with dat either it or bridge no longer has.

I think that it arises when one uses a particular Trezor device with multiple computers, potentially with those computers using multiple hardware wallets.  Sometimes, an error "can't find session" was displayed in the browser window after a long (>2min) timeout -- indicating that Bridge was working correctly, but the Trezor felt it had some kind of active session with another computer it could no longer pair with.  I don't know why MyCrypto sometimes works -- it might have to do with how they talk to the browser and bridge.

- For the MetaMask team - this is where I'd look to debug.  Assume the device is has multiple active sessions and might be getting confused -- MyCrypto might be instructing bridge to create a new one and that's why it's working.
- For the Trezor team - I think it might be good to look into this with both MyCrypto and MetaMask teams, so that if this error happens again in the future you get a specific error message for it, maybe with the specific device/session the Trezor is looking before it times out.  One device I tested with had 10+ sessions and none of them applied to any of the 2 computers we tried, and the Trezor had never been connected to any computers but those. This tells us that it's losing sessions?

In sum - if you use multiple hardware wallets on multiple computers - try using a completely fresh install of all touchpoints (Bridge, Browser + MetaMask, not just reinstalling MetaMask or clearing browser history).  I have been able to get this fixed for multiple people who never had Brave installed on their computer, just by installing Brave and getting the Trezor to talk to it. 

Otherwise (users with #1), carry on - it's probably best to wait until they get it all working and release a new version.

Stay safe out there. Standard disclaimer when commenting on wallet stuff in public forums: try these suggestions at your own risk, I am not responsible if your stuff breaks even more as a result of trying this, etc...

**Comment by HD125823** [comment]
@sdtsui I had a similar problem with being stuck on the loading page when connecting the trezor wallet and also got around this by clearing the cache (like you mentioned). Texted this https://github.com/MetaMask/metamask-extension/issues/8349#issuecomment-702332962 but no reply yet :/

**Comment by yohaaan** [comment]
@whymarrh hi, any update on this? Trezor users keep facing this issue. Thanks!

**Comment by unicornioPT** [comment]
6 months later and this has not yet been upgraded?! Unbelievable... It's just trezor wallet right... No big deal...

**Comment by vondra** [comment]
+1 solving the issue. 

Message from Trezor support: “Now it is up to MetaMask developers to update Trezor Connect to the latest version. We have already started the task by creating the issue in their GitHub repository. You can follow the progress under the following link: https://github.com/MetaMask/metamask-extension/issues/8349

You can alternatively downgrade your firmware to version 2.3.0 which is compatible with older Trezor Connect. Please see our user manual: https://wiki.trezor.io/Firmware_downgrade
”

**Comment by blockomoco** [comment]
Can someone from Metamask share status pls. @whymarrh @danfinlay

**Comment by Prince-Fish** [comment]
Again another week of no update. My displeasure and disappointment with metamask at this point is immense.  At this point the only thing I can do is vote with my feet and use another wallet. 

6 months to update a damn integration that has been proven to work across other platforms (completed by much smaller and less sophisticated teams). Shameful.

**Comment by unicornioPT** [comment]
> @epop-cs as @ncsolar and @dpazdan have shared, these do all seem to be related.
> 
> @blockomoco thanks for the details, that is clearer.
> 
> We'll be working to get this in 8.1, the version after next, and will update here with progress. We understand that this isn't a smooth process right now, we're working to fix it.

this did not age well

**Comment by HD125823** [comment]
@danfinlay Can we please get an update on this?

**Comment by blockomoco** [comment]
Looping in devs who have been active in this repo recently: @tmashuang @brad-decker @danjm 

You guys pls take a look and help resolving the issue.

Thank you and cheers.

**Comment by chonghe1** [comment]
Hope the issue can be resolved soon

**Comment by blockomoco** [comment]
When reflecting on this ticket, the lack of responsibility (and maybe even capability) is stunning. Im going to stop using Trezor as a consequence. Fairly speaking, the responsibility lies primarily in Metamask's dev team. Although Im not sure if the Trezor team could have been more engaged to resolve this issue (?). Don't care any more to be honest.

Because Metamask is so wide spread among Dapps, Im currently forced to keep on using it. Moving forward only with Ledger.

Unfollowing this thread. Good luck to you all and feel sorry for the users of Metamask + Trezor.

**Comment by gruvin** [comment]
fwiw, imo, it is Trezor that have let us all down. They were the ones who made the breaking changes. They know Metamask is as popular as it is and that it support(ed) Trezor. It is Trezor's product that is losing out ...

> Im going to stop using Trezor as a consequence. 

Just exactly how Metamask (or anyone else) is supposed to write code to allow entry of passphrase using only buttons on the Trezor 1 is beyond me and frankly laughable. No one would use it because it would be a horrible user experience.

I have three Trezor 1's and I've already ditched them all. I recently bought a $$$$ Ledger X (BT) but I can't stand the thing. It's horrible by comparison and the expen$ive BlueTooth feature DOESN'T WORK on desktops! So screw them too.

**Metmask is awesome**, especially on Mobile. Hardware Wallets seem to pretty much all suck the proverbial. Paper cold wallets are FAR superior for storing large sums of crypto anyway.

</two-cents>

**Comment by alangrainger** [comment]
@gruvin it's not a breaking change - Metamask + Trezor still functions fine, it's just that Metamask hasn't updated to allow Model T users to be able to enter the phrase on their device (unlike other web wallet integrations which have updated and do allow this).

I don't have a One, but I believe this feature is just for the T, so I'm not aware that there would be additional code to write - just an upgrade of the `trezor-connect` dependency.

**Comment by alangrainger** [comment]
@gruvin it's not a breaking change - Metamask + Trezor still functions fine, it's just that Metamask hasn't updated to allow Model T users to be able to enter the phrase on their device (unlike other web wallet integrations which have updated and do allow this).

I don't have a One, but I believe this feature is just for the T, so I'm not aware that there would be additional code to write - just an upgrade of the `trezor-connect` dependency..... which really seems like it shouldn't have taken 8 months an counting.

If any dev wants to get this done by the end of the year, I'll throw in a US$500 bounty in the crypto of your choice.

**Comment by alangrainger** [comment]
@gruvin it's not a breaking change - Metamask + Trezor still functions fine, it's just that Metamask hasn't updated to allow Model T users to be able to enter the phrase on their device (unlike other web wallet integrations which have updated and do allow this).

I don't have a One, but I believe this feature is just for the T, so I'm not aware that there would be additional code to write - just an upgrade of the `trezor-connect` dependency..... which really seems like it shouldn't have taken 8 months an counting.

**If any dev wants to get this done by the end of the year, I'll throw in a US$500 bounty in the crypto of your choice.**

**Comment by alangrainger** [comment]
@gruvin I don't have a One, but I believe this feature is just for the T, so I'm not aware that there would be additional UI code to write - just an upgrade of the `trezor-connect` dependency..... which really seems like it shouldn't have taken 8 months an counting.

**If any dev wants to get this done by the end of the year, I'll throw in a US$500 bounty in the crypto of your choice.**

**Comment by yohaaan** [comment]
> When reflecting on this ticket, the lack of responsibility (and maybe even capability) is stunning. Im going to stop using Trezor as a consequence. Fairly speaking, the responsibility lies primarily in Metamask's dev team. Although Im not sure if the Trezor team could have been more engaged to resolve this issue (?). Don't care any more to be honest.
> 
> Because Metamask is so wide spread among Dapps, Im currently forced to keep on using it. Moving forward only with Ledger.
> 
> Unfollowing this thread. Good luck to you all and feel sorry for the users of Metamask + Trezor.

We as Trezor has informed Metamask regarding the change in timely manner and also we do keep reminding them (this issue was created by us as well), unfortunately with no reaction from Metamask so far. You can use Trezor with other ETH Wallets including our native ETH Wallet available at https://suite.trezor.io/

**Comment by unicornioPT** [comment]
Out of curiosity... How many lines of code/hours are we talking about here?
This does not seem to be such a big upgrade.

**Comment by yohaaan** [comment]
@unicornioPT I cannot tell really as I'm not familiar with Metamask app (as we do not develop it) but other services have been able to adopt it really quickly, it shouldn't be that big of a change. We've also offered Metamask help (should they be facing any issue with implementation on their side) so the ball is really on their side.

**Comment by unicornioPT** [comment]
The odd thing is that a while ago metamask devs seemed ready to complete
the update.
Leaves me thinking if there are other issues in play here besides code.

**Comment by knd** [comment]
looks like there's no point following up on this. 

high chance there is internal conflicts/ issues within Metamask team that could not get this prioritized or worked on given quite number of user feedbacks.

**Comment by gruvin** [comment]
@alangrainger Howdy comrade.

All in good fun ... I'm in the mood for a <rant>:p

Well see, actually, *Trezor* themselves literally labelled the change, "breaking change". Trezor 1 also comes up with the scary warning about how Metamask [the API calling party] is no longer compatible and things could go wrong. The reason it's not compatible is a blanket version check of the API version being called.

That said, sure. In all actuality, the Trezor 1 does _seem_ to still work, so long as the scary warning message doesn't put users off. It does me. shrug

In the end I maintain that **Trezor** should submit a PR to Metamask to have the problem resolved. **Trezor's wallet. Trezor's customers. Trezor's "fault". Trezor's responsibility** [Given that this is Metmask and not just some random app. It deserves more respect.] At the very least, they could not show the damned warning for Trezor 1 devices, since imo it's basically moot. (No way to enter passphrase on a Trezor 1 anyhow.)

As things stand, it appears this issue aint never gonna get [properly] fixed. sigh
</rant>

Strange. I thought I'd feel better, having gotten that off my chest, again. Huh. Oh well.

**Comment by chonghe1** [comment]
Really hope this issue can be resolved soon. Many other apps like Mycrypto etc has updated it to support Trezor

**Comment by stx233** [comment]
Yes please update the Trezor library.

**Comment by unicornioPT** [comment]
I mentioned this before... The way metamask delt with this makes me think
there is politics involved... Their team was on it and it seemed to be a
quick and easy update then all of a sudden everything stopped.

**Comment by gruvin** [comment]
@unicornioPT seems plausible. Open source, community / transient-dev projects don't usually operate  that way though. It's just as likely that there simply aren't any devs on the present MM team who happen to own, use or care about Trezor hardware.

Trezor broke it. Trezor should fix it.

If they don't, frankly Metamask would be well within their collective rights to simply remove support for Trezor wallets entirely. There are _mkany_ other, not so well marketed hard wallets out there. Metamask are under no obligation to support any of them.

**Comment by unicornioPT** [comment]
While trezor and mm devs keep playing the blame game, we, users of both
products get less than we hope and paid for

**Comment by gruvin** [comment]
Indeed we do. I paid for my Trezors. I paid nothing for the 100% free Metamask. I blame Trezor. You should too. I would suggest we continue bitching at them, over at their site. They aint reading this, here, I'm sure.

**Comment by steelbom** [comment]
I'm sure Trezor made changes to their software for a reason. Other platforms have quickly updated to the latest version; Trezor can't do it on Metamask's behalf.

Of course, Metamask is not obligated to update (or continue supporting Trezor) - they can do whatever they want, but I assume there is a userbase of Trezor users and it is important to us Trezor users that they make the necessary changes so we can continue using Metamask **securely**.

**Comment by gruvin** [comment]
> there is a user base of Trezor users and it is **important** 

apparently not. sigh

**Comment by blockomoco** [comment]
Just visited this issue again, just for entertainment. 

Painful to see no progress whatsoever. And no reaction from the dev team...
Just reached out to the Consensys discord to raise awareness on this issue. Lets see if that helps.

**Comment by gruvin** [comment]
> Just reached out to the Consensys discord to raise awareness on this issue. Lets see if that helps.
thanks

**Comment by blockomoco** [comment]
There is a comment that suggests downgrading the Trezor firmware to 2.3.0 and that this resolve the reported compatibility issues. 

I tested this today and can confirm that downgrading to 2.3.0 does **not** help. The same issue exists as with more recent firmwares and that is: one needs to type the passphrase on the computer as opposed to type it on the device for signing transactions via Metamask.

**Comment by chonghe1** [comment]
> There is a comment that suggests downgrading the Trezor firmware to 2.3.0 and that this resolve the reported compatibility issues.
> 
> I tested this today and can confirm that downgrading to 2.3.0 does **not** help. The same issue exists as with more recent firmwares and that is: one needs to type the passphrase on the computer as opposed to type it on the device for signing transactions via Metamask.

No I didn't. But downgrading to 1.8.0 only can work. But it is a really old firmware it will erase the seed, and on top of that, 1.8.0 will have the annoying webusb notification everytime Trezor connects,

**Comment by yohaaan** [comment]
@gruvin sir this issue has been created by Trezor CTO, we were pro-actively offering help and provided Metamask with info about the change in a timely manner, there is no reply from their side, unfortunately. You are correct, you did pay for Trezor and Trezor device and Trezor apps are working fine, you can use our new app for storing ETH and ERC20 tokens - https://suite.trezor.io/

TBH this is like blaming Ford for Continental tires not being compatible with Ford cars, we cannot ensure that something we do not develop is not working.

Also there are two firmware versions, each one for one device model Trezor One has firmware versions starting with "1", Trezor T - starting with "2".

In general, if you are experiencing issues with Metamask we suggest using the native ETH/ERC20 wallet insteady.

**Comment by alangrainger** [comment]
@yohaaan in case you're not aware of the necessity for this integration - it's not so that people can store ERC20 tokens, it's so that we can use our Trezor to safely interact with Defi services.

I want to use Uniswap - can't currently do that with the Trezor since my passphrase would be exposed.

I want to lend on Fulcrum - can't do it.

I want to participate in almost **anything** in Defi using my Trezor..... and right now I'm not able to safely do it without exposing my passphrase.

🔥 **Again, I repeat my offer of a US$500 bounty to get this done.** 🔥

**Comment by unicornioPT** [comment]
Can I use trezor's native wallet to interact with web3 dapps?

**Comment by unicornioPT** [comment]
+1
These references to use the native wallet are hilarious... Wake up! Users
need mm to interact with DeFi dapps!
And BTW, trezor keeps stating they reached out to mm in a timely manner...
Guess that needs improvement.

FYI: mm seems to be very active in the consensys discord. Better luck
creating awareness to this bug there.

Small note...i have not seen any dialog on the discord between trezor and
mm, only unhappy customers.

**Comment by yohaaan** [comment]
@unicornioPT this is their official GitHub, it's common for devs to communicate via GitHub but thanks for the tip, I'll try to reach them over Discord to see what's going on.

**Comment by vjsingh-lite** [comment]
Ended up in this thread since I saw the same notification (new Trezor user and tried using with Metamask for the first time) and wanted to know more about this situation.

Honestly, I don't understand why integrating the newer version of Trezor connect should take so much time by Metamask. No dev replies over here is concerning too. It's a public repo. Of course, this should be one of the official places for issue communication.

**Comment by snazzybytes** [comment]
@vjsingh-lite @unicornioPT  I agree with all of you!! As we all know it is CRITICAL to not use laptop keyboard for entering passphrase. And anyone involved with crypto industry shall make such critical updates a priority especially for just small handful of OG veteran hardware wallets like Trezor/Ledger/KeepKey i guess. this is whack experience, but I will NOT be buying Ledger as i like Trezor for its open-source vibes and touch screen. I just simply will not use MetaMask then (because I am not sending one more transaction from Trezor to MetaMask browser wallet address :D thats more money in fees than necessary). Then MetaMask is for complete newbs who will become victims of scams sooner rather than later and hardcore users who plug in USB to confirm will disappear.  someone should blow this situation up on crypto Twitter to get it on the radar and get it done under more pressure.

@danfinlay any progress on your fix? @yohaaan stated that Trezor team is willing to help with any technical issues that may arise during the programming session - that seems more than one can ask for - willingness to help with software engineering challenge just like any fortune-100s/200s would do when integrating APIs together that benefit both sides. what is the hold up then?

**Comment by snazzybytes** [comment]
@whymarrh @danfinlay @yohaaan 
would this PR be of any help to get the ball rolling? 
disclaimer, i have no clue about this codebase, just simply bumping versions and see if the tests pass (some failed, adjusted the expected mock data to get them to pass).

**Comment by stx233** [comment]
@snazzybytes Thanks for trying to make it done !

**Comment by snazzybytes** [comment]
@stx233 yea no problem, sadly i do not have tribal knowledge to actually get it properly done obviously (i work as software engineer, but on the backend side of things with  different set of technologies JVM and Golang and other things and languages, not the NPM and javascript on the daily though as is the case here, sorry)

@stx233 also good sign is Jacob from MetaMask added the "S1-High" label below your reply, which is a good sign it will be prioritized and squashed soon I hope I'd imagine logically 👨‍🚀 speaking (tagging you twice, in case you notice the good news sooner, cuz it certainly put a smile on my face)

**Comment by danfinlay** [comment]
@yohaaan wrote:
> sir this issue has been created by Trezor CTO, we were pro-actively offering help and provided Metamask with info about the change in a timely manner, there is no reply from their side, unfortunately.

This is inaccurate. The MetaMask official replies have ironically become buried by people asking where our replies are. I'm going to lock the thread so that my response here can be more easily found, and we'll be providing updates here as we develop this issue.

We had a developer spend a week on this integration a couple months ago, but ran into some issues introduced by the new Trezor library update that we were not able to overcome with the time allotted to it. We will be spending more time tracking these issues down soon, but needed to focus on some wider-impact improvements, like transaction reliability.

Here were the outputs of his efforts:
 -------
The problem is that when trying to connect Trezor, we are redirected to https://connect.trezor.io/8/popup.html#loading and it’s getting stuck on loading screen (exactly the same way when we manually open this site), there’s nothing more we can do. I have latest firmware on my model T and latest trezor-bridge versions.
I tried updating trezor-connect to many different 8.x.x versions (including extended one), but none of them works - extended versions runs some unit tests, but should resolve if we have a public key test is getting timeout from TrezorConnecty.getPublicKey() function. Also tried using .ethereumGetPublicKey() , but it behaves in the same way
Same function works fine using https://trezor.github.io/connect-explorer/#/
I tested same function using their example from https://github.com/trezor/connect/tree/develop/examples/webextension and it also works so it looks like the problem of implementation, not firmware/bridge versions
I also checked how MEW has this implemented and it also looks almost the same :man-shrugging:
I have no clue what is going on, so any tips or suggestions will be appreciated.
 -------

Those efforts are on some slightly stale branches, so we're going to bring them up to date before sharing again here, where we're happy to get input from the Trezor team.

We'd also love to have more direct communication with the Trezor team, maybe we could get a shared Slack or Discord going like we have with Ledger, so we can interact a bit more tightly. Please do reach out.

Again, locking this thread as our official replies keep getting buried with wild speculation.

**Comment by darkwing** [comment]
@prusnak Thank you for reporting this issue and driving us forward!

After a few days of debugging, I've found the same issue that the previous engineer found (https://github.com/MetaMask/metamask-extension/issues/8349#issuecomment-754323808).

Essentially, the `TrezorConnect.getPublicKey` call never resolves and the user indefinitely sees the loading spinner at https://connect.trezor.io/8/popup.html#loading .  

![TrezorConnect](https://user-images.githubusercontent.com/46655/104132924-eda82280-5345-11eb-9832-cc02d4752ca6.gif)

No error or warning is reported in the JS console, nor does a promise catch -- just infinite loading screen.

The MetaMask code that calls the `TrezorConnect.getPublicKey` is in another repository; I've pushed a PR with the update to `"trezor-connect@extended": "^8.1.19"` here:  https://github.com/MetaMask/eth-trezor-keyring/pull/46

A community-submitted PR, https://github.com/MetaMask/metamask-extension/pull/10131 , effectively shims MetaMask to use `"trezor-connect@extended": "^8.1.19"` which allows you to see the MetaMask-to-Trezor connect flow.  Since both repositories show `TrezorConnect.getPublicKey` stalling, I don't believe it's a specific MetaMask integration issue.

Any advice would be appreciated.  This is my top priority and I'll be looking to act as quickly as your team can help.  Thank you!

**Comment by prusnak** [comment]
@darkwing Are you 100% sure there is no browser extension messing up with the communication? (privacy badger, ad block, etc.?)

Pinging @szymonlesisz our Connect guru: Szymon, any ideas, what might be wrong with the Connect integration used by Metamask?

**Comment by darkwing** [comment]
I read the changelog for 8.0.0 (https://github.com/trezor/connect/blob/develop/CHANGELOG.md#800) and the first point mentioned a communication change but if postMessage is a fallback, I would think everything should work properly on that front, since version 7 has worked for us without problems.

I look forward to any advice from @szymonlesisz ; one question to consider is what could cause the connect website to infinitely show a loading screen.  Thank you!

**Comment by Gudahtt** [comment]
I've just been testing with https://github.com/MetaMask/eth-trezor-keyring/pull/46 on Chrome (Version 87.0.4280.88 (Official Build) (64-bit)) on Ubuntu, by linking that version of `eth-trezor-keyring` with the `develop` branch of the extension. I'm using an otherwise empty Chrome profile, so no blockers. I'm seeing the exact same thing as @darkwing.

**Comment by darkwing** [comment]
Today we found that there was a seemingly undocumented breaking change to Trezor Connect 8 that, once we fixed, will allow Trezor to work on that version:

https://gist.github.com/darkwing/b68125069533628ea2c27445cb253b99

I'm working to complete the upgrade in https://github.com/MetaMask/eth-trezor-keyring/pull/46; the only current blocker is that our test suite's `TrezorConnect.getPublicKey` test hangs because that method no longer throws an error when called.  In previous versions, it would throw a missing `document` error, but with the `extended` package, that's no longer the case.  I'm working now to figure out how to fix our tests.

**Comment by blockomoco** [comment]
@darkwing Which Metamask browser version is going to include this change ?

**Comment by shaimo** [comment]
Any eta for this? I’m still unable to connect my Trezor...

**Comment by chonghe1** [comment]
> Any eta for this? I’m still unable to connect my Trezor...

Hey the latest Metamask already solved this, v9.0.4. 

PS: Thanks to the team, very glad to be able to use Trezor with Metamask now

**Comment by shaimo** [comment]
I’m using v9.0.4 and it doesn’t work for me. Gave up on it...

**Comment by chonghe1** [comment]
> I’m using v9.0.4 and it doesn’t work for me. Gave up on it...

It really should work. To be clear, we are talking about the issue of inserting passphrase on Trezor T device, not computer. 

Do you mean you cannot connect Trezor?

**Comment by shaimo** [comment]
I simply can’t connect Metamask to Trezor. It opens a popup for a second in another tab and then disappears, and then I get an ever going spinning wheel and that’s it. Tried it with two different trezors, and in both Brave and Chrome. Always the same

**Comment by Gudahtt** [comment]
@shaimo you're likely experiencing a different problem than what is described in this issue. Could you create a separate issue, and provide any details you can to help us reproduce the problem?

**Comment by shaimo** [comment]
@Gudahtt https://github.com/MetaMask/metamask-extension/issues/10372
Thanks!

**Comment by stx233** [comment]
Version 9.0.4 is okay. 

It recognized trezor model T and we can use it with metamask.

But there is an issue. Here is what I found : 
- Sending BNB from bsc wallet to another bsc wallet on trezor T with metamask : Ok no problem
- Sending from Trezor T wallet with metamask using bsc to another bsc wallet : it does not work and displays transaction failed (I tried several times, with different browsers FF, chrome, brave)

how can I display the output message of  in order to help you ?
